cli/src/commands/git/colocation.rs

```diff
@@ -0,0 +1,343 @@
+    if git::get_git_backend(workspace_command.repo().store()).is_err() {
+        return Err(user_error(
+            "This command requires a repository backed by Git. This repository appears to be \
+             using a different backend.",
+        ));
+    }
+
+    // Ensure that this is the main workspace
+    let workspace_root = workspace_command.workspace_root();
+    let jj_repo_path = workspace_root.join(".jj").join("repo");
+    if jj_repo_path.is_file() {
+        return Err(user_error(
+            "This command cannot be used in a non-main Jujutsu workspace.",
+        ));
+    }
+
+    match subcommand {
+        GitColocationCommand::Enable(args) => cmd_git_colocation_enable(ui, command, args),
+        GitColocationCommand::Disable(args) => cmd_git_colocation_disable(ui, command, args),
+        GitColocationCommand::Status(args) => cmd_git_colocation_status(ui, command, args),
+    }
+}
+
+fn cmd_git_colocation_status(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    _args: &GitColocationStatusArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+    let is_colocated =
+        is_colocated_git_workspace(workspace_command.workspace(), workspace_command.repo());
+
+    if is_colocated {
+        writeln!(ui.stdout(), "Repository is currently colocated with Git.")?;
+        writeln!(
+            ui.hint_default(),
+            "To disable colocation, run: `jj git colocation disable`"
+        )?;
+    } else {
+        writeln!(
+            ui.stdout(),
+            "Repository is currently not colocated with Git."
+        )?;
+        writeln!(
+            ui.hint_default(),
+            "To enable colocation, run: `jj git colocation enable`"
+        )?;
+    }
+
+    Ok(())
+}
+
+fn cmd_git_colocation_enable(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    _args: &GitColocationEnableArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+
+    // Check if the repo is already colocated before proceeding
+    if is_colocated_git_workspace(workspace_command.workspace(), workspace_command.repo()) {
+        writeln!(ui.status(), "Repository is already colocated with Git.")?;
+        return Ok(());
+    }
+
+    let workspace_root = workspace_command.workspace_root();
+    let dot_jj_path = workspace_root.join(".jj");
+    let jj_repo_path = dot_jj_path.join("repo");
+    let git_store_path = jj_repo_path.join("store").join("git");
+    let git_target_path = jj_repo_path.join("store").join("git_target");
+    let dot_git_path = workspace_root.join(".git");
+
+    // This is called after checking that the repo is backed by git, is not
+    // already colocated, and is a main workspace, but we must still bail out
+    // if a git repo already exist at the root folder
+    if dot_git_path.exists() {
+        return Err(user_error(
+            "A .git directory already exists in the workspace root. Cannot colocate.",
+        ));
```

Comment:
Do we need to check this again? Perhaps, we can instead do `fs::rename()` first, which would fail if `.git` directory existed for unknown reasons.

Reply:
This check is not done before this point. At this point we know that the repo is git backed and not colocated, but there could still be an existing .git folder at the root of the repository. I don't like just letting rename fail because the error message will be very generic. I think this should stay as it is.